add language name to kernelspec

minrk opened the pull request:
This should be the same for any kernel from a given project, indicating compatibility. If you don't have a kernel that matches exact name, a kernel with a matching project_name will be selected.

If no matching kernel is found, a dialog is shown:

![screen shot 2015-01-14 at 10 34 39](https://cloud.githubusercontent.com/assets/151929/5744840/f9646aa0-9bd8-11e4-9242-2d58999e09f0.PNG)

closes #7349

minrk committed:
add project_name to kernelspec

this should be the same for compatible kernels ('ipython', 'ijulia', etc.)

minrk committed:
handle missing kernel

- if project_name matches, pick another kernel
- if no match, show dialog with kernel selector

takluyver commented:
The dialog looks like it assumes that you should be starting some kernel. I think it also needs to be easy to open a notebook with no kernel at all, when you just don't have the relevant kernel installed. E.g. if I open an Haskell notebook, none of the kernels available are a suitable substitute.

minrk commented:
@takluyver I think working well without a kernel is not in scope for 3.0. Code everywhere assumes there is a running kernel, and I don't think we can remove that reliably in our current timeframe.

minrk commented:
That said, I can add a 'cancel' button that enters the 'Dead kernel' state, which indicates that "things are broken", which is accurate.

takluyver commented:
Yes, something like that. I would call it something more precise like 'Don't start a kernel' instead of 'Cancel'.

minrk commented:
An alternative to project_name would be putting language name and version in the kernelspec. I know we moved that information out of the kernelspec because language info kept growing, but it may make sense to have one or two keys of redundant info in the kernelspec to address this compatibility issue.

minrk commented:
project_name doesn't provide any mechanism for a notebook to 'require' Python 3. I'm not sure we need to support that, but project_name doesn't have any of the relevant information.

minrk commented:
Forgot to ping @fperez and @ellisonbg for the missing kernel behavior

fperez commented:
ping, Travis failure (I didn't look at the log)

minrk commented:
Yup, I'll finish it up after we've talked about it tomorrow.

ellisonbg commented:
Trying to understand exactly what this does...

When I originally proposed the "project_name" idea, I was thinking of it as a human readable name that we could use in the notebook UI (such as an on hover to give more information about the kernel) to help users understand more about the kernel. This would allow us to keep the main visible name short and focused on the language, while allowing a user to dig deeper into the actual project providing that language.

Sounds like you are using the project name in a different way to handle certain types of errors? Can you describe a bit more of the motivation?

minrk commented:
I wasn't addressing the text on hover at all, I'm not sure how or why that would differ from display_name. This is purely meant to address the issue that the *name* of the kernelspec doesn't address the compatibility issue.

For example: I have 6 IPython kernels right now. If I create a notebook and associate it with my 'python3-debug' kernel, you should still be able to open it, as long as you have *any* IPython kernel.

An alternative could be to copy some subset of the language_info we get from the kernel_info request back into the kernelspec (where it started), and use that to identify compatibility.

We could also punt on this question entirely, and only keep the dialog used when no *identical* kernel can be found. That would mean that any user running Python 2 would get a dialog saying "I couldn't find a Python 3 kernel" whenever they open an example notebook.

ellisonbg commented:
Ahh, Ok thanks for the explanation. Now I understand this PR much better.

Hmm, it seems like tackling the whole kernel compatibility thing in an automated way is a bit crazy. While there are definitely some obvious cases where there is a clear answer, there are others where it will be difficult. Here is one proposal:

* If a kernel for a notebook doesn't exist, open the notebook and let the user pick which installed kernel they want to use. If they don't have any kernel then they can pick the best one and at least edit the notebook and not run code.
* Just use the project name as I described above - for human usage in on hover events to learn more about a kernel so if there are 2 kernels with the display name "Python 2" they can see i) the name of the kernel and ii) the name of the project.

Another idea would be for us to implement and provide a "null" kernel that does nothing, but allows a users to edit a notebook. But it would at least have all of the right keyboard events hooks up to make cell execution do nothing in a good way.

Carreau commented:
> and provide a "null" kernel that does nothing, but allows a users to edit a notebook

You mean a in-browser js kernel ?

ellisonbg commented:
Originally I wasn't thinking a JS based kernel, but that probably makes the most sense. It is a different way of handling the "no kernel" case that may lead to simpler code.

minrk commented:
I think leaving it in the dead-kernel state makes the most sense for 3.0. Writing a new dummy kernel doesn't seem reasonable at this point, and most of the severe UI problems when the kernel is dead or missing have been addressed.

fperez commented:
+1
On Jan 16, 2015 5:47 PM, "Min RK" <notifications@github.com> wrote:

> I think leaving it in the dead-kernel state makes the most sense for 3.0.
> Writing a new dummy kernel doesn't seem reasonable at this point, and most
> of the severe UI problems when the kernel is dead or missing have been
> addressed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipython/pull/7454#issuecomment-70349244>.
>

takluyver commented:
From the dev meeting:

* We will do something like this, but with a field called language_name instead of project_name.
* First it will check for an exact match with the kernel name
* If there isn't a match, it will find all kernels with matching language name
  * If there's only one, start that
  * If there's more than one, prompt the user to select from those that match
  * If there are no matches, prompt the user to select from any available kernels, with a prominent option to not start a kernel.

minrk committed:
replace project_name with language_name

- automatically choose language match, if only one
- only list language matches in select dialog
- list all kernels if no language match

minrk committed:
add "No appropriate kernel" button to kernel-selection prompt

minrk committed:
s/language_name/language/

ellisonbg commented:
@minrk is this ready for review?

minrk commented:
Updated per the dev meeting, this now uses a `language` field, instead of `project_name`, which is `python` for any Python kernel.

1. if kernelspec name matches, use it (no change from master)
2. if exactly one kernel's language matches, use it (no dialog)
3. otherwise, prompt with selection from:
    - matches, if more than one
    - all kernels, if no matches

A "No appropriate kernel" button allows the user to use the notebook without starting a kernel.

The dialog:

![screen shot 2015-01-20 at 15 28 40](https://cloud.githubusercontent.com/assets/151929/5828449/18530586-a0b9-11e4-9a69-e8396f7bd1f2.PNG)

short list:

![screen shot 2015-01-20 at 15 28 05](https://cloud.githubusercontent.com/assets/151929/5828456/27891dce-a0b9-11e4-9bac-d1a883b4595b.PNG)

long list:

![screen shot 2015-01-20 at 15 25 42](https://cloud.githubusercontent.com/assets/151929/5828459/2e62075a-a0b9-11e4-8db6-b13c5bc45794.PNG)

minrk commented:
@ellisonbg as of a moment ago, it should be.

minrk committed:
add language key to kernelspec doc

minrk commented:
Should be ready to go.

Carreau commented:
The "No Appropriate Kernel" seem more like a statement than a action to do for me. 
If you find a better sentence, like "Continue without kernel" or whatever, that would be good. 
Thoughts ?

mathieu1 commented:
Maybe this has been discussed already, but is there a plan to display a warning in the second case (no match for kernelspec name but exactly one kernel's language matches), maybe in the terminal that lauched the notebook server?

Carreau commented:
> Maybe this has been discussed already, but is there a plan to display a warning in the second case (no match for kernelspec name but exactly one kernel's language matches), maybe in the terminal that lauched the notebook server?

Would be complex to do in the terminal.

In the notification area maybe ?

ellisonbg commented:
I agree on the "Continue without kernel" idea.

On Fri, Jan 23, 2015 at 12:39 AM, Matthias Bussonnier <
notifications@github.com> wrote:

> The "No Appropriate Kernel" seem more like a statement than a action to do
> for me.
> If you find a better sentence, like "Continue without kernel" or whatever,
> that would be good.
> Thoughts ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipython/pull/7454#issuecomment-71162864>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

minrk commented:
@Carreau I added a notification message, but it's immediately dismissed because there are other notifications that follow it as part of kernel startup. Any way to get a message that will stick without being clobbered by subsequent messages? If not, I'd be inclined to leave out the notification, for now.

minrk committed:
use Array.map

minrk committed:
"Continue without kernel" button

minrk commented:
Button renamed to "Continue without kernel".

Carreau commented:
> @Carreau I added a notification message, but it's immediately dismissed because there are other notifications that follow it as part of kernel startup. Any way to get a message that will stick without being clobbered by subsequent messages? If not, I'd be inclined to leave out the notification, for now.

I think that should be possible, let me check.

Carreau commented:
There use to be a one shot notification widget (which click on it would destroy), but it was not used so remove. 

Do you want me to redo it with a convenience methods. Something like `notification_area.one_shot_warning(message)` ?

minrk commented:
@Carreau create a widget, display the message, and destroy on dismiss? I think I can handle that.

minrk committed:
use one-off notification widget for match notification

when no exact match, and only one kernel matches language

minrk commented:
one-off notification used (there was already a method that worked fine). The text is "Using kernel: <display_name>", and it's info-level. I don't think warning is appropriate. A more detailed message is logged, as well.

takluyver commented:
UI nitpicking - if the dialog still looks like this:

![screenshot](https://cloud.githubusercontent.com/assets/151929/5828449/18530586-a0b9-11e4-9a69-e8396f7bd1f2.PNG)

Can we have a little bit of space between the colon and the dropdown? It looks ugly to have them pressed so close together.

takluyver commented on IPython/html/static/notebook/js/kernelselector.js:
```diff
@@ -154,28 +160,116 @@ define([
+            };
+        }
+        if (this._loaded) {
+            this._set_kernel(selected);
```
What's the rationale behind having this extra branch for if it's already loaded? I thought promises encapsulate the same logic, so just doing `this.loaded.then(...)` should be equivalent.

I suspect the answer involves async somehow, but I'd like to see the explanation. :horse:

minrk replied:
This can be called multiple times, so if I keep calling promise.then, the callback will fire N times, rather than once.

takluyver commented on IPython/html/static/notebook/js/kernelselector.js:
```diff
@@ -154,28 +160,116 @@ define([
+            }
+            // if still undefined, trigger failure event
+            if (ks === undefined) {
+                if (matches) {}
```
Should there be some code inside this `if`? It corresponds to finding multiple kernels matching the language field.

minrk replied:
No, nothing here. The case is handled later. Removed this dead bit.

takluyver commented on IPython/html/static/notebook/js/kernelselector.js:
```diff
@@ -154,28 +160,116 @@ define([
+        names.map(function (name) {
+            var ks = that.kernelspecs[name];
+            select.append(
+                $('<option/>').attr('value', ks.name).text(ks.spec.display_name || ks.name)
```
Why is display_name nested under a `spec` namespace, while name isn't?

minrk replied:
Because display_name is part of a kernelspec, the spec's name (the folder in which it's contained) is not part of the spec itself.

minrk committed:
add some horizontal margin on select dropdown

to avoid overlapping adjacent text when active

minrk commented:
Space around the dropdown added:

![screen shot 2015-01-26 at 15 27 49](https://cloud.githubusercontent.com/assets/151929/5910589/973ecb8a-a570-11e4-96c2-2e72d5422ec1.PNG)

takluyver commented on IPython/html/static/notebook/js/kernelselector.js:
```diff
@@ -154,28 +160,115 @@ define([
+            }
+            // if still undefined, trigger failure event
+            if (ks === undefined) {
+                this.events.trigger("spec_not_found.Kernel", {
```
OK, so the `spec_not_found` event really represents two distinct conditions: multiple matches were found, and no matches were found. I'd be inclined to have separate events for those, and present the dialogs slightly differently to make it clearer what's going on. If we do keep a single event, I think it probably warrants an explanatory comment that it can result from two cases.

minrk replied:
Since it's handled exactly the same other than the content of the selector, I'm not sure they should be separate events. I guess they could be two events with the same handler.
